feat(assets): Add Support for Electrical Component Connections

eduardiazf opened the pull request:
### Summary
Add a new client method to retrieve microgrid electrical components connections. For this a new ComponentConnection class is introduced. The CLI is also extended to add a command to list component-connections, for example: `assets-cli component-connections 123 --source 5 --source 6 --destination 10`



#### Resources
- [https://github.com/frequenz-floss/frequenz-client-microgrid-python/blob/v0.18.x/src/frequenz/client/microgrid/component/_connection.py](https://github.com/frequenz-floss/frequenz-client-microgrid-python/blob/v0.18.x/src/frequenz/client/microgrid/component/_connection.py)

dependabot[bot] committed:
Bump actions/labeler from 5.0.0 to 6.0.1

Bumps [actions/labeler](https://github.com/actions/labeler) from 5.0.0 to 6.0.1.
- [Release notes](https://github.com/actions/labeler/releases)
- [Commits](https://github.com/actions/labeler/compare/8558fd74291d67161a8a78ce36a881fa63b766a9...634933edcd8ababfe52f92936142cc22ac488b1b)

---
updated-dependencies:
- dependency-name: actions/labeler
  dependency-version: 6.0.1
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
feat(assets): List electrical component connections

Add a new client method to retrieve microgrid electrical components connections.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
feat(tests): Tests electrical component connections

Adds tests for list_microgrid_electrical_component_connections, covering success, empty results, and error cases to improve reliability.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -138,3 +137,47 @@ async def list_microgrid_electrical_components(
+        self,
+        microgrid_id: int,
+        source_component_ids: list[int] | None = None,
+        destination_component_ids: list[int] | None = None,
```
You could use this instead to make it more generic and even simplify the code:


```suggestion
        source_component_ids: Iterable[int] = (),
        destination_component_ids: Iterable[int] = (),
```

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -138,3 +137,47 @@ async def list_microgrid_electrical_components(
+            request.source_component_ids.extend(source_component_ids)
+
+        if destination_component_ids:
+            request.destination_component_ids.extend(destination_component_ids)
```
Then you don't need to check for `None`:

```suggestion
        request = assets_pb2.ListMicrogridElectricalComponentConnectionsRequest(
            microgrid_id=microgrid_id,
            source_component_ids.extend=source_component_ids,
            destination_component_ids=destination_component_ids,
        )
```

llucax reviewed (commented):
LGTM, but I would improve the API as proposed. With this change not only it can accept `list`s of IDs, it can also accept tuples, sets, or any iterable, like `dict.values()` or `dict.keys()`.

eduardiazf committed:
refactor(assets): Update list_microgrid_electrical_component_connections

Updates list_microgrid_electrical_component_connections to accept Iterable[int] for source/destination component IDs.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

cwasicki commented on src/frequenz/client/assets/_client.py:
```diff
@@ -138,3 +139,43 @@ async def list_microgrid_electrical_components(
+
+        return [
+            component_connection_from_proto(connection)
+            for connection in response.connections
```
IIUC this would also collect `None`s, is this intended?

llucax replied:
True, good catch. I have this in the microgrid API:

```py
        return (
            conn
            for conn in map(
                component_connection_from_proto,
                response.electrical_component_connections,
            )
            if conn is not None
        )
```

An a more functional alternative :smile: 

```py
        return list(
            map(
                component_connection_from_proto,
                filter(bool, response.electrical_component_connections),
            )
        )
```

eduardiazf replied:
Thanks, good catch.

cwasicki commented on src/frequenz/client/assets/electrical_component/_connection_proto.py:
```diff
@@ -0,0 +1,106 @@
+            )
+    else:
+        minor_issues.append(
+            "missing operational lifetime, considering it always operational",
```
Is this an issue or a valid case?

llucax replied:
In the microgrid API is valid, and for now it is always returned without a lifetime, but maybe this needs to change in the future, not sure.

We should probably document in the protobuf files what to expect though. Some of these issues I added them only to be notified if they happened or not because I couldn't find a strict definition.

Once we have a clear definition, we might want to remove them (for example, if it is a defined thing that a missing operational lifetime can happen and then the meaning is it is always operational, we don't need to be whining about it when we see it happen, it will be just noise.

@tiyash-basu-frequenz FYI

tiyash-basu-frequenz replied:
This is already documented in the latest API version: https://github.com/frequenz-floss/frequenz-api-common/blob/300bcb4296d679faa3b2aec33e7ea3d461d90beb/proto/frequenz/api/common/v1alpha8/microgrid/lifetime.proto#L21-L30

Therefore, the phrase "missing operational lifetime, considering it always operational" is technically correct. I do not think it should be classified as an issue though.

ktickner replied:
is there a reason it is this concise? It's nice to  reduce cognitive load of messages, but this is verging on being easily misunderstood.

llucax replied:
> Therefore, the phrase "missing operational lifetime, considering it always operational" is technically correct. I do not think it should be classified as an issue though.

Yeah, it is is expected behavior it is not an issue and we should remove it, as it will just spam the logs. I will remove it from the microgrid API client too.

> is there a reason it is this concise? It's nice to reduce cognitive load of messages, but this is verging on being easily misunderstood.

What are you talking about exactly, what's concise?

Marenz replied:
My guess: He means the message `"missing operational lifetime, considering it always operational"` is concise and easily misunderstood and could use maybe a tiny bit more context

ktickner replied:
yes thank you @Marenz

cwasicki commented on src/frequenz/client/assets/electrical_component/_connection_proto.py:
```diff
@@ -0,0 +1,106 @@
+        except ValueError as exc:
+            major_issues.append(
+                f"invalid operational lifetime ({exc}), considering it as missing "
+                "(i.e. always operational)",
```
Do we want to consider it as operational? If so, I don't see a way how we can detect these cases programatically. Maybe we add a way to check how many major issues were found to understand how reliable the connections are.

llucax replied:
> Do we want to consider it as operational?

Good question, this was done with the microgrid API in mind, where right now everything is operational all the time basically, but maybe it is not safe to assume it operational.

> If so, I don't see a way how we can detect these cases programatically. Maybe we add a way to check how many major issues were found to understand how reliable the connections are.

The question for me is if there is anything we can do (and want to do) programatically. I think in general an error here is a misconfiguration, and the best we can do is tell the user (in this case an error should be logged) so they can fix it.

Then we probably need to try to recover as best as possible, so there is the question if we ignore the connection (consider it non-operational) or we add it (assume it is operational). Both cases will be wrong some times and right some other times, so I guess there is no good answer for this, but maybe ignoring it is more correct, as if one field is wrong, maybe other fields are wrong too.

Or back to your question, what else do you think it can be done by a client if the wrong connection would be passed to the user of the client? Do you envision clients applying some heuristics to try to do a better guess on how to treat a faulty connection?

tiyash-basu-frequenz replied:
See [here](https://github.com/frequenz-floss/frequenz-client-assets-python/pull/50#discussion_r2469426160).

cwasicki replied:
I agree that it makes sense to define missing as always operational. What I am not sure about is whether an invalid lifetime should be accepted and treated as missing. Invalid indicates a user error to me.

cwasicki replied:
> Or back to your question, what else do you think it can be done by a client if the wrong connection would be passed to the user of the client? Do you envision clients applying some heuristics to try to do a better guess on how to treat a faulty connection?

I guess my main point would be to have a way to validate whether the graph could be parsed without any issues or not, because atm we only have a warning message. It depends on the use-case if you prefer to have reliable data or better some data. Also you could try to recover it or have fallbacks in place if a problem was detected.

llucax replied:
I think the graph will be constructed from the UI, and there will be no way to construct it in an invalid way, so I guess this code is mostly about really weird situations, memory corruption or something like this, so I think as long as we log an error (I think major issues are logged as warnings now, we might want to reconsider that, or have a separate `critical_issues` that are logged as errors), it doesn't matter much how we deal with it.

llucax replied:
That said, probably if part of a message is broken it would make sense to ignore the whole message instead try to guess fix it.

llucax replied:
I think you removed the wrong "issue", this issue should still be checked and reported (I think it would probably be better to just skip the whole connection object, so return `None`, if the operational lifetime is **invalid**. If it is missing, then that's OK, so that issue should not be reported.

cwasicki replied:
If there is no supported way to construct invalid graphs and it's only weird cases, are these then valid at all? If not, why not just raise an exception in this case?

llucax replied:
I didn't understand the last comment. Maybe we could meet to quickly discuss this.

eduardiazf committed:
refactor(assets): Refactors list_microgrid_electrical_component_connections

To use map/filter for cleaner logic and to skip invalid connections.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

Marenz commented:
Hmm a bit meta but it I found it a bit cumbersome to read every time "This commit introduced/updates/.."  it feels very redundant. Nothing against AI generated if that is the case, but maybe ask it to cut it down a bit? :)

llucax commented:
> Hmm a bit meta but it I found it a bit cumbersome to read every time "This commit introduced/updates/.." it feels very redundant. Nothing against AI generated if that is the case, but maybe ask it to cut it down a bit? :)

I send this, I also find the PR description too verbose, to the point that my brain unconsciously tries to skip it. For me something more succinct would be more useful, like:

> Add a new client method to retrieve microgrid electrical components connections. For this a new `ComponentConnection` class is introduced. The CLI is also extended to add a command to list component-connections, for example: `assets-cli component-connections 123 --source 5 --source 6 --destination 10`

I think the rest of the details can be read from the PR itself.

Maybe you can tweak the instructions to generate PRs/commit messages, which makes me think that maybe it is time to start providing custom instructions, as I guess we can all benefit from having good instructions to produce useful PR/commit messages.

dependabot[bot] committed:
Bump the compatible group with 2 updates

Bumps the compatible group with 2 updates: [frequenz-floss/gh-action-nox](https://github.com/frequenz-floss/gh-action-nox) and [frequenz-floss/gh-action-setup-python-with-deps](https://github.com/frequenz-floss/gh-action-setup-python-with-deps).

Updates `frequenz-floss/gh-action-nox` from 1.0.0 to 1.0.1
- [Release notes](https://github.com/frequenz-floss/gh-action-nox/releases)
- [Commits](https://github.com/frequenz-floss/gh-action-nox/compare/v1.0.0...v1.0.1)

Updates `frequenz-floss/gh-action-setup-python-with-deps` from 1.0.0 to 1.0.1
- [Release notes](https://github.com/frequenz-floss/gh-action-setup-python-with-deps/releases)
- [Commits](https://github.com/frequenz-floss/gh-action-setup-python-with-deps/compare/v1.0.0...v1.0.1)

---
updated-dependencies:
- dependency-name: frequenz-floss/gh-action-nox
  dependency-version: 1.0.1
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: compatible
- dependency-name: frequenz-floss/gh-action-setup-python-with-deps
  dependency-version: 1.0.1
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: compatible
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
refactor(assets): Operational lifetime
missing operational lifetime, considering it always operational

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

llucax commented on src/frequenz/client/assets/cli/__main__.py:
```diff
@@ -201,6 +205,97 @@ async def list_microgrid_electrical_components(
+                    list(destination_component_ids)
+                    if destination_component_ids
+                    else None
+                ),
```
Why converting to `list`? The client method takes `Iterator` so it should work without the conversion, right?

eduardiazf replied:
Yes, that's right, we can omit the conversion.

llucax commented on src/frequenz/client/assets/electrical_component/_connection_proto.py:
```diff
@@ -0,0 +1,96 @@
+        return lifetime_from_proto(message.operational_lifetime)
+    minor_issues.append(
+        "missing operational lifetime, considering it always operational",
+    )
```
We wanted to remove this issue reporting as a missing operational lifetime is supported officially, so it is OK.

```suggestion
```

llucax commented on src/frequenz/client/assets/electrical_component/_connection.py:
```diff
@@ -0,0 +1,74 @@
+    """
+
+    operational_lifetime: Lifetime | None = None
+    """The operational lifetime of the connection."""
```
This should not accept `None`, otherwise we need to do special casing here, is better to keep it as it was before, that a `Lifetime()` is just always operational. Also in user code, if users need to access operational lifetime directly they will need to deal with `None` as a possibility now.
```suggestion
    operational_lifetime: Lifetime
    """The operational lifetime of the connection."""
```

llucax commented on src/frequenz/client/assets/electrical_component/_connection.py:
```diff
@@ -0,0 +1,74 @@
+    def is_operational_at(self, timestamp: datetime) -> bool:
+        """Check whether this connection is operational at a specific timestamp."""
+        if self.operational_lifetime is None:
+            return True
```
```suggestion
```

llucax commented on src/frequenz/client/assets/electrical_component/_connection_proto.py:
```diff
@@ -0,0 +1,93 @@
+    """Get the operational lifetime from a protobuf message."""
+    if message.HasField("operational_lifetime"):
+        return lifetime_from_proto(message.operational_lifetime)
+    return None
```
```suggestion
    return Lifetime()
```

cwasicki commented on src/frequenz/client/assets/electrical_component/_connection_proto.py:
```diff
@@ -0,0 +1,93 @@
+    """Create a `ComponentConnection` from a protobuf message collecting issues.
+
+    This function is useful when you want to collect issues during the parsing
+    of multiple connections, rather than logging them immediately.
```
What is actually the motivation for collecting those messages instead of logging them immediately? I wonder why we need this additional complexity.

llucax replied:
Flexibility. Doing this actually makes supporting your use case pretty trivial. The client could just return the issues instead of logging them. This code is supposed to go to client-common.

cwasicki replied:
If this is the idea, then these issues are  probably better tracked via enums instead of strings. If you prefer to not do it in this PR then it should come in the next one. Because in the current form we cannot use this endpoint in our trading app.

eduardiazf replied:
We can add it in the next one because this implementation should be moved to client-common.

cwasicki replied:
Sounds good!

llucax replied:
> If this is the idea, then these issues are probably better tracked via enums instead of strings.

Yeah, this could be a good idea, or just exceptions.

But for now I would do the minimal changes we need to support your use case (so I would stay with strings for now).

Then I would prioritize moving this to `api-common` so any improvements we made we can use in all APIs using components and connections and we avoid diverging implementations in different clients.

llucax reviewed (approved):
LGTM. I suggest to merge this as is and after we agree on an approach to support @cwasicki use case, we implement #53 in a separate PR.
